lib/mapObjects/CGHeroInstance.cpp

```diff
@@ -1412,29 +1412,28 @@ std::vector<SecondarySkill> CGHeroInstance::getLevelupSkillCandidates(IGameRando
 			basicAndAdv.insert(elem.first);
 		none.erase(elem.first);
 	}
-
-	if (!basicAndAdv.empty())
+	
+	for(;;)
 	{
-		skills.push_back(gameRandomizer.rollSecondarySkillForLevelup(this, basicAndAdv));
-		basicAndAdv.erase(skills.back());
-	}
+		if(skills.size() >= cb->getSettings().getInteger(EGameSettings::HEROES_SKILL_SELECTION_AMOUNT_UPGR) || basicAndAdv.empty())
+			break;
```

Comment:
This should be a simple while loop instead of infinite loop with conditional break:

(ugh, github blocks suggestion mode since 1 line is "unchanged")
```cpp
int maxUpgradedSkills = cb->getSettings().getInteger(EGameSettings::HEROES_SKILL_SELECTION_AMOUNT_UPGR);

while (skills.size() < maxUpgradedSkills && !basicAndAdv.empty())
{
	skills.push_back(gameRandomizer.rollSecondarySkillForLevelup(this, basicAndAdv));
	basicAndAdv.erase(skills.back());
}
```